### ISSUE TITLE
[BUG]: when passing test_data , metrics scores do not make sense

### ISSUE THREAD

**Issue by atwahsz** [pycaret/pycaret#3083]
### pycaret version checks

- [X] I have checked that this issue has not already been reported [here](https://github.com/pycaret/pycaret/issues).

- [X] I have confirmed this bug exists on the [latest version](https://github.com/pycaret/pycaret/releases) of pycaret.

- [X] I have confirmed this bug exists on the master branch of pycaret (pip install -U git+https://github.com/pycaret/pycaret.git@master).

### Issue Description

when splitting the data before hand , and setting up with test_data  and then try to fit a model to the data , the final mean metrics do not represent the real metrics when applying the model to the test_data

![image](https://user-images.githubusercontent.com/24771158/201467342-eef6983e-fa63-4b1a-b44e-6f82ad47cce9.png)

as u can see  the mean r2 of 10 kfold is 0.0551 , while the prediction r2 in the testing data is 0.9847.

### Reproducible Example

```python
from pycaret.datasets import get_data
from pycaret.regression import *
house = get_data('house')
train=house.head(int(len(house)*0.7))
test=house.tail(int(len(house)*0.2))
setup(data=train,test_data=test,target='SalePrice',use_gpu=True)
xgb= create_model('xgboost')
test1 = predict_model(xgb,test)
```

### Expected Behavior

the resulting r2 should be equal of that of the prediction if test_data is passed

### Actual Results

```python-traceback
as shown in the image above
```

### Installed Versions

'3.0.0.rc4'

**Comment by moezali1** [comment]
@atwahsz This seems like specific to this dataset. I have tried fitting a linear model and don't see such a gap between train and test set. Xgboost for some reasons is not doing a good job on this dataset. Here is a residual plot from xgboost:

![image](https://user-images.githubusercontent.com/54699234/202852816-934e35c6-76da-43cd-b622-bdc586d5aa51.png)

Also, I noticed the way you sampled `train` and `test` set might create bias. I suggest if you want to sample using `head` and `len` method, you should shuffle the afterwards before passing it to `setup`.

see the results from `compare_models`. all linear models are top performing and tree based model below 0.15 R2.

![image](https://user-images.githubusercontent.com/54699234/202852885-7f8fd222-d1e3-4afe-809e-9caf58a5cef1.png)

**Comment by moezali1** [comment]
@tvdboom I have run the house dataset through 2.3.10 and also 3.0rc4. I see wildly different cross-validation results. I am more concerned about that rather than differences between train and test scores. I believe these huge swings in performance are because of difference in preprocessing pipelines between 2 and 3. I am concerned that it could be some kind of bug in the pipeline of 3.0. Would you mind taking a look?

## PyCaret 3.0-rc4

```
from pycaret.datasets import get_data
data = get_data('house')

from pycaret.regression import *
s = setup(data, target = 'SalePrice', session_id = 123, ignore_features = ['Id'])

best = compare_models()
```
![image](https://user-images.githubusercontent.com/54699234/202853428-1720ae8b-d436-4021-b8a7-0d3903b7383c.png)

## PyCaret 2.3.10

```
from pycaret.datasets import get_data
data = get_data('house')

from pycaret.regression import *
s = setup(data, target = 'SalePrice', session_id = 123, ignore_features = ['Id'], silent = True)

best = compare_models()
```

![image](https://user-images.githubusercontent.com/54699234/202853331-f6a42bf9-eeb4-496f-8447-7ef76fa9e57b.png)

**Comment by atwahsz** [comment]
See how LightGBM preforming differently

**Comment by tvdboom** [comment]
@moezali1 This is not a bug. The difference is that 3.0 doesn't use one hot encoding for categorical columns with more than 5 classes (default value `max_encoding_ohe=5`). When set to a higher value, 3.0 performs better than 2.3.10.

![image](https://user-images.githubusercontent.com/32366550/203071179-1e02375c-535f-4b50-b059-88610f11ec1d.png)

I changed the default value of max_encoding_ohe to 10 in #3101. That's probably a more sensible number.

**Comment by moezali1** [comment]
@tvdboom What do you think about keeping the default same as 2.X and have this as a feature?

**Comment by tvdboom** [comment]
Leaving it to unlimited can cause unexpected memory issues for columns with many categorical values. If you think 10 is too few I would suggest to set it higher, maybe 20 or 25

**Comment by moezali1** [comment]
Ok @tvdboom let's set it to 25 then and close this issue once done.

**Comment by tvdboom** [comment]
ok. I changed it in #3101
